static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,382 @@
+): StandardMessage<{ width: number; height: number }> => ({
+	id: 'aaaa0000-bb11-cc22-dd33-eeeeee444444',
+	type: 'resize',
+	iframeId: payload.value.id,
+	slotId: payload.value.slotId,
+	value: {
+		height: +payload.value.height,
+		width: +payload.value.width,
+	},
+});
+
+/**
+ * Retrieve a reference to the calling iframe via it's ID.
+ *
+ * Incoming messages contain the ID of the iframe into which the source window is embedded.
+ */
+const getIframe = (data: StandardMessage) => {
+	if (data.slotId) {
+		const container = document.getElementById(`dfp-ad--${data.slotId}`);
+		const iframes = container
+			? container.getElementsByTagName('iframe')
+			: null;
+		return iframes?.length ? iframes[0] : null;
+	} else if (data.iframeId) {
+		const el = document.getElementById(data.iframeId);
+		if (el instanceof HTMLIFrameElement) {
+			return el;
+		}
+		return null;
+	}
+};
+
+/**
+ * Convert an `unknown` payload to the standard message format
+ *
+ * Until DFP provides a way for us to identify with 100% certainty our
+ * in-house creatives, we are left with doing some basic tests
+ * such as validating the anatomy of the payload and whitelisting
+ * event type
+ */
+const isValidPayload = (payload: unknown): payload is StandardMessage => {
+	const payloadToCheck = payload as StandardMessage;
+	return (
+		'type' in payloadToCheck &&
+		'value' in payloadToCheck &&
+		'id' in payloadToCheck &&
+		payloadToCheck.type in LISTENERS &&
+		/^[a-f0-9]{8}-([a-f0-9]{4}-){3}[a-f0-9]{12}$/.test(payloadToCheck.id)
+	);
+};
+
+/**
+ * Cheap string formatting function
+ *
+ * @param error An object `{ code, message }`. `message` is a string where successive
+ * occurrences of %% will be replaced by the following arguments
+ * @param args Arguments that will replace %%
+ *
+ * @example
+ * formatError({ message: "%%, you are so %%" }, "Regis", "lovely")
+ * => { message: "Regis, you are so lovely" }
+ */
+const formatError = (error: { message: string }, ...args: string[]) =>
+	args.reduce((e, arg) => {
+		e.message = e.message.replace('%%', arg);
+		return e;
+	}, error);
+
+/**
+ * Convert a posted message to our StandardMessage format
+ *
+ * @param event The message event received on the window
+ * @returns A message with the `StandardMessage` format, or null if the conversion was unsuccessful
+ */
+const eventToStandardMessage = (event: MessageEvent) => {
+	try {
+		// Currently all non-string messages are discarded here since parsing throws an error
+		// TODO Review whether this is the desired outcome
+		const data: unknown = JSON.parse(event.data);
+		return isProgrammaticMessage(data)
+			? toStandardMessage(data)
+			: isValidPayload(data)
+			? data
+			: null;
+	} catch (ex) {
+		return null;
+	}
+};
+
+/**
+ * Callback that is fired when an arbitrary message is received on the window
+ *
+ * @param event The message event received on the window
+ */
+const onMessage = (
+	event: MessageEvent<string>,
+): Promise<unknown> | undefined => {
```

Comment:
I think you can turn this into an `async` function that always return a promise. It’s hard to reason about something that may return a promise or not.

```suggestion
const onMessage = async (
	event: MessageEvent<string>,
): Promise<void> => {
```

Reply:
Good shout!